base/task.jl

```diff
@@ -365,6 +365,73 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    tasks = Task[]
+    done_mask = Bool[]
+    exception = false
+    nremaining::Int = 0
+
+    for (i, t) in enumerate(waiting_tasks)
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+        if istaskdone(t)
+            push!(done_mask, true)
+            exception |= istaskfailed(t)
+        else
+            push!(done_mask, false)
+            nremaining += 1
+        end
+    end
+
+    if nremaining == 0
+        return tasks, Task[]
+    elseif any(done_mask) && (!all || (failfast && exception))
+        return tasks[done_mask], tasks[.~done_mask]
+    end
+
+    chan = Channel{Tuple{Int,Task}}(Inf)
+    waiter_tasks = Dict{Task, Task}()
+
+    for (i, done) in enumerate(done_mask)
+        if !done
```

Comment:
performance nit: add a check here to break if something finishes early?
```suggestion
        done && continue
        if istaskdone(t)
            t = tasks[i]
            exception |= istaskfailed(t)
            done_mask[i] = true
            exception && failfast && break
        else
```